<a href="https://colab.research.google.com/github/LyaSolis/clinical-trial-information-extraction/blob/main/2_Extract_Paragraphs.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
from google.colab import drive
drive.mount('/content/drive/')

Drive already mounted at /content/drive/; to attempt to forcibly remount, call drive.mount("/content/drive/", force_remount=True).


In [2]:
!ls drive/MyDrive/LAG/Articles_Parsed

images	pdf_files  temp  xml_files


### Drop Corrupt Files (by size)

In [3]:
%%bash
cd drive/MyDrive/LAG/Articles_Parsed/
mkdir temp

mkdir: cannot create directory ‘temp’: File exists


In [28]:
%%bash
cd drive/MyDrive/LAG/Articles_Parsed/xml_files
find . -type f -size -11k

./j.1399-6576.1998.tb05330.x.cermxml
./00000542-199404000-00007.cermxml
./j.cnki.cjap.2016.05.006.cermxml
./00003246-199301000-00017.cermxml
./s-2008-1067758.cermxml
./00024382-199711000-00003.cermxml
./bja%2F64.4.493.cermxml
./S2213-2600%2820%2930394-5.cermxml
./00000542-198606000-00018.cermxml
./09031936.01.17100080.cermxml
./00003246-198907000-00003.cermxml
./00004872-198811000-00008.cermxml
./NEJM199112123252404.cermxml
./00005344-199404000-00001.cermxml
./a%3A1010604913656.cermxml
./s-2003-40371.cermxml
./00004872-198610000-00010.cermxml
./00003246-199406000-00014.cermxml
./chest.128.2.714.cermxml
./s-2008-1068292.cermxml
./00006231-199903000-00007.cermxml
./NEJM199302113280605.cermxml
./00005344-199105000-00007.cermxml


In [29]:
%%bash
cd drive/MyDrive/LAG/Articles_Parsed/xml_files
find . -type f -size -11k -exec mv "{}" ../temp/ \;

In [4]:
!ls -lh drive/MyDrive/LAG/Articles_Parsed/temp

total 63K
-rw------- 1 root root 1.4K Jun 10 09:40 00000542-198606000-00018.cermxml
-rw------- 1 root root 2.5K Jun 10 09:12 00000542-199404000-00007.cermxml
-rw------- 1 root root 1.6K Jun 10 09:46 00003246-198907000-00003.cermxml
-rw------- 1 root root 1.6K Jun 10 09:14 00003246-199301000-00017.cermxml
-rw------- 1 root root 1.4K Jun 10 10:29 00003246-199406000-00014.cermxml
-rw------- 1 root root 1.8K Jun 10 10:22 00004872-198610000-00010.cermxml
-rw------- 1 root root 2.1K Jun 10 09:53 00004872-198811000-00008.cermxml
-rw------- 1 root root 1.8K Jun 10 11:30 00005344-199105000-00007.cermxml
-rw------- 1 root root 2.1K Jun 10 10:05 00005344-199404000-00001.cermxml
-rw------- 1 root root 1.8K Jun 10 11:06 00006231-199903000-00007.cermxml
-rw------- 1 root root 1.8K Jun 10 09:22 00024382-199711000-00003.cermxml
-rw------- 1 root root 5.4K Jun 10 09:45 09031936.01.17100080.cermxml
-rw------- 1 root root 3.2K Jun 10 10:05 a%3A1010604913656.cermxml
-rw------- 1 root root 4.3K Jun 10 09:2

In [5]:
!cat drive/MyDrive/LAG/Articles_Parsed/temp/00000542-199404000-00007.cermxml

<!DOCTYPE article PUBLIC "-//NLM//DTD JATS (Z39.96) Journal Archiving and Interchange DTD v1.0 20120330//EN" "JATS-archivearticle1.dtd">
<article xmlns:xlink="http://www.w3.org/1999/xlink">
  <front>
    <journal-meta />
    <article-meta />
  </front>
  <body>
    <fig>
      <graphic xlink:href="Articles/00000542-199404000-00007.images/img_1_1.png" />
    </fig>
    <fig>
      <graphic xlink:href="Articles/00000542-199404000-00007.images/img_1_2.png" />
    </fig>
    <fig>
      <graphic xlink:href="Articles/00000542-199404000-00007.images/img_2_1.png" />
    </fig>
    <fig>
      <graphic xlink:href="Articles/00000542-199404000-00007.images/img_2_2.png" />
    </fig>
    <fig>
      <graphic xlink:href="Articles/00000542-199404000-00007.images/img_3_1.png" />
    </fig>
    <fig>
      <graphic xlink:href="Articles/00000542-199404000-00007.images/img_3_2.png" />
    </fig>
    <fig>
      <graphic xlink:href="Articles/00000542-199404000-00007.images/img_4_1.png" />
    </fig>
   

In [6]:
!cat drive/MyDrive/LAG/Articles_Parsed/temp/S2213-2600%2820%2930394-5.cermxml

<!DOCTYPE article PUBLIC "-//NLM//DTD JATS (Z39.96) Journal Archiving and Interchange DTD v1.0 20120330//EN" "JATS-archivearticle1.dtd">
<article xmlns:xlink="http://www.w3.org/1999/xlink">
  <front>
    <journal-meta>
      <journal-title-group>
        <journal-title>Published online August</journal-title>
      </journal-title-group>
    </journal-meta>
    <article-meta>
      <title-group>
        <article-title>Pharmacotherapy for pulmonary arterial hypertension</article-title>
      </title-group>
      <pub-date>
        <year>2020</year>
      </pub-date>
      <volume>25</volume>
      <issue>2020</issue>
    </article-meta>
  </front>
  <body>
    <sec id="sec-1">
      <title>-</title>
      <p>A new study has concluded that There was no significant diference
adding selexipag to macitentan and in change in PVR between the
tadalafil does not improve pulmonary two groups. The patients given
vascular resistance in patients with double combination therapy saw an
pulmonary arter

In [7]:
%%bash
cd drive/MyDrive/LAG/Articles_Parsed/xml_files
ls | wc -l

836


836 Files left for further work

### Parse Files

Testing it on a smaller file (incorrectly parsed, but suitable for testing code)

In [8]:
with open("drive/MyDrive/LAG/Articles_Parsed/temp/S2213-2600%2820%2930394-5.cermxml", "r") as f:
  file_content = f.readlines()
fl = []
for line in file_content:
  fl.append(line)
fl

['<!DOCTYPE article PUBLIC "-//NLM//DTD JATS (Z39.96) Journal Archiving and Interchange DTD v1.0 20120330//EN" "JATS-archivearticle1.dtd">\n',
 '<article xmlns:xlink="http://www.w3.org/1999/xlink">\n',
 '  <front>\n',
 '    <journal-meta>\n',
 '      <journal-title-group>\n',
 '        <journal-title>Published online August</journal-title>\n',
 '      </journal-title-group>\n',
 '    </journal-meta>\n',
 '    <article-meta>\n',
 '      <title-group>\n',
 '        <article-title>Pharmacotherapy for pulmonary arterial hypertension</article-title>\n',
 '      </title-group>\n',
 '      <pub-date>\n',
 '        <year>2020</year>\n',
 '      </pub-date>\n',
 '      <volume>25</volume>\n',
 '      <issue>2020</issue>\n',
 '    </article-meta>\n',
 '  </front>\n',
 '  <body>\n',
 '    <sec id="sec-1">\n',
 '      <title>-</title>\n',
 '      <p>A new study has concluded that There was no significant diference\n',
 'adding selexipag to macitentan and in change in PVR between the\n',
 'tadalafi

In [9]:
paragraph_start = []
paragraph_end = []

for ind, el in enumerate(fl):
  if "<p>" in el:
    p_start = ind
    paragraph_start.append(p_start)
  if "</p>" in el:
    p_end = ind +1
    paragraph_end.append(p_end)


for st, end in zip(paragraph_start, paragraph_end):
  print(st, end)


22 32
32 58
58 70


In [10]:
paragraphs = []
for st, end in zip(paragraph_start, paragraph_end):
  paragraphs.append(fl[st:end])

paragraphs

[['      <p>A new study has concluded that There was no significant diference\n',
  'adding selexipag to macitentan and in change in PVR between the\n',
  'tadalafil does not improve pulmonary two groups. The patients given\n',
  'vascular resistance in patients with double combination therapy saw an\n',
  'pulmonary arterial hypertension improvement in PVR of 52%, relative\n',
  '(PAH). The results were presented at to baseline levels, whereas those\n',
  'this year’s American Thoracic Society assigned the triple combination\n',
  'International Conference, which was saw an improvement of 54%. “The\n',
  'held online because of the COVID-19 results were fairly consistent across\n',
  'pandemic. subgroups”, said Kelly Chin (University</p>\n'],
 ['      <p>The investigators randomly assigned of Texas Southwestern Medical\n',
  '247 patients with PAH to receive first- Center, Dallas, TX, USA), the study’s\n',
  'line treatment either with macitentan corresponding author. Secondary\n',
  

Wrapping it into a single cell

In [11]:
fl = []
with open("drive/MyDrive/LAG/Articles_Parsed/temp/S2213-2600%2820%2930394-5.cermxml", "r") as f:
  file_content = f.readlines()
  for line in file_content:
    fl.append(line)


paragraph_start = []
paragraph_end = []

for ind, el in enumerate(fl):
  if "<p>" in el:
    p_start = ind
    paragraph_start.append(p_start)
  if "</p>" in el:
    p_end = ind +1
    paragraph_end.append(p_end)


paragraphs = []
for st, end in zip(paragraph_start, paragraph_end):
  paragraphs.append(fl[st:end])

clean_paragraphs = []
for paragraph in paragraphs:
  clean_paragraph = []
  for line in paragraph:
    line = line.strip()
    line = line.replace('<p>', '')
    line = line.replace('</p>', '')
    clean_paragraph.append(line)
  clean_paragraphs.append(clean_paragraph)

clean_paragraphs

[['A new study has concluded that There was no significant diference',
  'adding selexipag to macitentan and in change in PVR between the',
  'tadalafil does not improve pulmonary two groups. The patients given',
  'vascular resistance in patients with double combination therapy saw an',
  'pulmonary arterial hypertension improvement in PVR of 52%, relative',
  '(PAH). The results were presented at to baseline levels, whereas those',
  'this year’s American Thoracic Society assigned the triple combination',
  'International Conference, which was saw an improvement of 54%. “The',
  'held online because of the COVID-19 results were fairly consistent across',
  'pandemic. subgroups”, said Kelly Chin (University'],
 ['The investigators randomly assigned of Texas Southwestern Medical',
  '247 patients with PAH to receive first- Center, Dallas, TX, USA), the study’s',
  'line treatment either with macitentan corresponding author. Secondary',
  'and tadalafil, followed by the placebo endpoint

In [12]:
with open('txt_file.txt', 'w') as x:
    x.write('\n'.join(' '.join(map(str, row)) for row in clean_paragraphs))


In [13]:
!cat txt_file.txt

A new study has concluded that There was no significant diference adding selexipag to macitentan and in change in PVR between the tadalafil does not improve pulmonary two groups. The patients given vascular resistance in patients with double combination therapy saw an pulmonary arterial hypertension improvement in PVR of 52%, relative (PAH). The results were presented at to baseline levels, whereas those this year’s American Thoracic Society assigned the triple combination International Conference, which was saw an improvement of 54%. “The held online because of the COVID-19 results were fairly consistent across pandemic. subgroups”, said Kelly Chin (University
The investigators randomly assigned of Texas Southwestern Medical 247 patients with PAH to receive first- Center, Dallas, TX, USA), the study’s line treatment either with macitentan corresponding author. Secondary and tadalafil, followed by the placebo endpoints were also similar. at 15 days (n=124), or macitentan “This study ad

Test it on a large file from dataset

In [14]:
%%bash
cd drive/MyDrive/LAG/Articles_Parsed/xml_files
find . -type f -size +180k

./1745-6215-14-91.cermxml


In [15]:
!head drive/MyDrive/LAG/Articles_Parsed/xml_files/1745-6215-14-91.cermxml

<!DOCTYPE article PUBLIC "-//NLM//DTD JATS (Z39.96) Journal Archiving and Interchange DTD v1.0 20120330//EN" "JATS-archivearticle1.dtd">
<article xmlns:xlink="http://www.w3.org/1999/xlink">
  <front>
    <journal-meta>
      <journal-title-group>
        <journal-title>Trials</journal-title>
      </journal-title-group>
    </journal-meta>
    <article-meta>
      <article-id pub-id-type="doi">10.1186/1745-6215-14-91</article-id>


In [28]:
fl = []
with open("drive/MyDrive/LAG/Articles_Parsed/xml_files/1745-6215-14-91.cermxml", "r") as f:
  file_content = f.readlines()
  for line in file_content:
    fl.append(line)


paragraph_start = []
paragraph_end = []

for ind, el in enumerate(fl):
  if "<p>" in el:
    p_start = ind
    paragraph_start.append(p_start)
  if "</p>" in el:
    p_end = ind +1
    paragraph_end.append(p_end)


paragraphs = []
for st, end in zip(paragraph_start, paragraph_end):
  paragraphs.append(fl[st:end])

clean_paragraphs = []
for paragraph in paragraphs:
  clean_paragraph = []
  for line in paragraph:
    line = line.strip()
    line = line.replace('<p>', '')
    line = line.replace('</p>', '')
    clean_paragraph.append(line)
  clean_paragraphs.append(clean_paragraph)

with open('txt_file.txt', 'w') as x:
    x.write('\n'.join(' '.join(map(str, row)) for row in clean_paragraphs))

In [29]:
!cat txt_file.txt

Background: Pulmonary arterial hypertension is a rare disorder associated with poor survival. Endothelial dysfunction plays a central role in the pathogenesis and progression of pulmonary arterial hypertension. Inflammation appears to drive this dysfunctional endothelial phenotype, propagating cycles of injury and repair in genetically susceptible patients with idiopathic and disease-associated pulmonary arterial hypertension. Therapy targeting pulmonary vascular inflammation to interrupt cycles of injury and repair and thereby delay or prevent right ventricular failure and death has not been tested. Spironolactone, a mineralocorticoid and androgen receptor antagonist, has been shown to improve endothelial function and reduce inflammation. Current management of patients with pulmonary arterial hypertension and symptoms of right heart failure includes use of mineralocorticoid receptor antagonists for their diuretic and natriuretic effects. We hypothesize that initiating spironolactone t

In [32]:
# Checking re naming file works:
output_dir = "drive/MyDrive/LAG/Articles_Parsed/temp/"
filename = "test.xml"
filename = (os.path.basename(filename).split('.')[0])
with open(os.path.join(output_dir, f'{filename}.txt'), 'w') as x:
    x.write('\n'.join(' '.join(map(str, row)) for row in clean_paragraphs))
  

In [34]:
!cat drive/MyDrive/LAG/Articles_Parsed/temp/test.txt

Background: Pulmonary arterial hypertension is a rare disorder associated with poor survival. Endothelial dysfunction plays a central role in the pathogenesis and progression of pulmonary arterial hypertension. Inflammation appears to drive this dysfunctional endothelial phenotype, propagating cycles of injury and repair in genetically susceptible patients with idiopathic and disease-associated pulmonary arterial hypertension. Therapy targeting pulmonary vascular inflammation to interrupt cycles of injury and repair and thereby delay or prevent right ventricular failure and death has not been tested. Spironolactone, a mineralocorticoid and androgen receptor antagonist, has been shown to improve endothelial function and reduce inflammation. Current management of patients with pulmonary arterial hypertension and symptoms of right heart failure includes use of mineralocorticoid receptor antagonists for their diuretic and natriuretic effects. We hypothesize that initiating spironolactone t

Loop over directory and save results

In [37]:
# Testing on discarded files:
import os
input_dir = "drive/MyDrive/LAG/Articles_Parsed/temp/"
output_dir = "drive/MyDrive/LAG/Articles_Parsed/temp/"
for filename in os.listdir(input_dir):
  fl = []
  with open(os.path.join(input_dir,filename), "r") as f:
    file_content = f.readlines()
    for line in file_content:
      fl.append(line)


  paragraph_start = []
  paragraph_end = []

  for ind, el in enumerate(fl):
    if "<p>" in el:
      p_start = ind
      paragraph_start.append(p_start)
    if "</p>" in el:
      p_end = ind +1
      paragraph_end.append(p_end)


  paragraphs = []
  for st, end in zip(paragraph_start, paragraph_end):
    paragraphs.append(fl[st:end])

  clean_paragraphs = []
  for paragraph in paragraphs:
    clean_paragraph = []
    for line in paragraph:
      line = line.strip()
      line = line.replace('<p>', '')
      line = line.replace('</p>', '')
      clean_paragraph.append(line)
    clean_paragraphs.append(clean_paragraph)
  filename = (os.path.basename(filename).split('.')[0])
  with open(os.path.join(output_dir, f'{filename}.txt'), 'w') as x:
    x.write('\n'.join(' '.join(map(str, row)) for row in clean_paragraphs))


In [38]:
!ls drive/MyDrive/LAG/Articles_Parsed/temp/

00000542-198606000-00018.cermxml  09031936.txt
00000542-198606000-00018.txt	  a%3A1010604913656.cermxml
00000542-199404000-00007.cermxml  a%3A1010604913656.txt
00000542-199404000-00007.txt	  bja%2F64.4.493.cermxml
00003246-198907000-00003.cermxml  bja%2F64.txt
00003246-198907000-00003.txt	  chest.128.2.714.cermxml
00003246-199301000-00017.cermxml  chest.txt
00003246-199301000-00017.txt	  j.1399-6576.1998.tb05330.x.cermxml
00003246-199406000-00014.cermxml  j.cnki.cjap.2016.05.006.cermxml
00003246-199406000-00014.txt	  j.txt
00004872-198610000-00010.cermxml  NEJM199112123252404.cermxml
00004872-198610000-00010.txt	  NEJM199112123252404.txt
00004872-198811000-00008.cermxml  NEJM199302113280605.cermxml
00004872-198811000-00008.txt	  NEJM199302113280605.txt
00005344-199105000-00007.cermxml  s-2003-40371.cermxml
00005344-199105000-00007.txt	  s-2003-40371.txt
00005344-199404000-00001.cermxml  s-2008-1067758.cermxml
00005344-199404000-00001.txt	  s-2008-1067758.txt
00006231-199903000-00007.ce

In [41]:
!cat drive/MyDrive/LAG/Articles_Parsed/temp/S2213-2600%2820%2930394-5.txt

A new study has concluded that There was no significant diference adding selexipag to macitentan and in change in PVR between the tadalafil does not improve pulmonary two groups. The patients given vascular resistance in patients with double combination therapy saw an pulmonary arterial hypertension improvement in PVR of 52%, relative (PAH). The results were presented at to baseline levels, whereas those this year’s American Thoracic Society assigned the triple combination International Conference, which was saw an improvement of 54%. “The held online because of the COVID-19 results were fairly consistent across pandemic. subgroups”, said Kelly Chin (University
The investigators randomly assigned of Texas Southwestern Medical 247 patients with PAH to receive first- Center, Dallas, TX, USA), the study’s line treatment either with macitentan corresponding author. Secondary and tadalafil, followed by the placebo endpoints were also similar. at 15 days (n=124), or macitentan “This study ad

### Running on dataset

In [53]:
!mkdir drive/MyDrive/LAG/Articles_Parsed/extracted_paragraphs
!ls drive/MyDrive/LAG/Articles_Parsed/

extracted_paragraphs  images  pdf_files  temp  xml_files  xml_files_bk


In [47]:
!cp -r drive/MyDrive/LAG/Articles_Parsed/xml_files drive/MyDrive/LAG/Articles_Parsed/xml_files_bk
!ls drive/MyDrive/LAG/Articles_Parsed/

extracted_paragraphs  images  pdf_files  temp  xml_files  xml_files_bk


In [51]:
filename = "j.1399-6576.1998.tb05330.x.cermxml"
print(os.path.basename(filename).split('.cermxml')[0])

j.1399-6576.1998.tb05330.x


In [54]:
import os
input_dir = "drive/MyDrive/LAG/Articles_Parsed/xml_files/"
output_dir = "drive/MyDrive/LAG/Articles_Parsed/extracted_paragraphs/"
for filename in os.listdir(input_dir):
  fl = []
  with open(os.path.join(input_dir,filename), "r") as f:
    file_content = f.readlines()
    for line in file_content:
      fl.append(line)


  paragraph_start = []
  paragraph_end = []

  for ind, el in enumerate(fl):
    if "<p>" in el:
      p_start = ind
      paragraph_start.append(p_start)
    if "</p>" in el:
      p_end = ind +1
      paragraph_end.append(p_end)


  paragraphs = []
  for st, end in zip(paragraph_start, paragraph_end):
    paragraphs.append(fl[st:end])

  clean_paragraphs = []
  for paragraph in paragraphs:
    clean_paragraph = []
    for line in paragraph:
      line = line.strip()
      line = line.replace('<p>', '')
      line = line.replace('</p>', '')
      clean_paragraph.append(line)
    clean_paragraphs.append(clean_paragraph)
  filename = (os.path.basename(filename).split('.cermxml')[0])
  with open(os.path.join(output_dir, f'{filename}.txt'), 'w') as x:
    x.write('\n'.join(' '.join(map(str, row)) for row in clean_paragraphs))


In [57]:
!cat drive/MyDrive/LAG/Articles_Parsed/extracted_paragraphs/s1053-0770%2805%2980140-5.txt


Objectives: It is customary to extubate patients immediately after closed mitral valvotomy. These patients often have deranged respiratory function caused by chronic lung congestion. The left ventricular function may also be subnormal after valvotomy in some patients. Therefore, elective ventilation for some duration in the postoperative period can be beneficial to these patients. This work is an attempt to find whether elective ventilation should be preferred over immediate extubation in these patients. Design: A prospective randomized study. Setting: The study was performed in a tertiary care hospital, and the patients are referred from the northern states of India. Participants: One hundred patients undergoing elective closed mitral valvotomy were included in the initial part of the study. Ten more patients were studied to evaluate the effect of mild hypercarbia on right ventricular function after closed mitral valvotomy. Interventions: One hundred patients were divided into t w o g